If no service matches a DNS request, return empty

blue-troy opened the pull request:
Write the dns message response body when it is not a ServerFailure

Coredns assumes ServeDNS has written the response body unless
the returned code indicates it hasn't been written as determined
by ClientWrite (which is only the codes RcodeServerFailure,
RcodeRefused, RcodeFormatError and RcodeNotImplemented),
in which case it writes the response.

If there's no next plugin, it simply returns the code but if
it isn't one of the ClientWrite codes, then no response is written.

If no response,client will try to resend dns request multiple times,
which causes seconds delay.

Fixes #859
Signed-off-by: blue-troy <12729455+blue-troy@users.noreply.github.com>

<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our developer guide: https://submariner.io/development/
2. Ensure you have added the appropriate tests for your PR: https://submariner.io/development/code-review/#test-new-functionality
3. Read the code review guide to ease the review process: https://submariner.io/development/code-review/
4. If the PR is unfinished, mark it as a draft: https://submariner.io/development/code-review/#mark-work-in-progress-prs-as-drafts
5. If you are using CI to debug, use your private fork: https://submariner.io/development/code-review/#use-private-forks-for-debugging-prs-by-running-ci
6. Add labels to the PR as appropriate.

This template is based on the K8s/K8s template:

https://github.com/kubernetes/kubernetes/blob/master/.github/PULL_REQUEST_TEMPLATE.md
-->

skitt reviewed (changes requested):
In your commit message, please describe the change you’re making, not just what it’s supposed to fix. Something like

```
If no service matches a DNS request, return empty

Chaining to the next server results in long delays before returning failure.
Since we know we manage all of .clusterset.local, return empty instead of chaining when no matching service is found.
```

(You’d have to check whether the “we know …” assumption is correct.)

blue-troy commented:
> 

thank for you remind,i will edit it

blue-troy commented:
in [coredns code](https://github.com/coredns/coredns/blob/master/plugin/kubernetes/handler.go#L65) , it handle NameError like below it return the RcodeSuccess and write code RcodeNameError to the Msg


```go
	if k.IsNameError(err) {
		if k.Fall.Through(state.Name()) {
			return plugin.NextOrFailure(k.Name(), k.Next, ctx, w, r)
		}
		if !k.APIConn.HasSynced() {
			// If we haven't synchronized with the kubernetes cluster, return server failure
			return plugin.BackendError(ctx, &k, zone, dns.RcodeServerFailure, state, nil /* err */, plugin.Options{})
		}
		return plugin.BackendError(ctx, &k, zone, dns.RcodeNameError, state, nil /* err */, plugin.Options{})
	}

func BackendError(ctx context.Context, b ServiceBackend, zone string, rcode int, state request.Request, err error, opt Options) (int, error) {
	m := new(dns.Msg)
	m.SetRcode(state.Req, rcode)
	m.Authoritative = true
	m.Ns, _ = SOA(ctx, b, zone, state, opt)

	state.W.WriteMsg(m)
	// Return success as the rcode to signal we have written to the client.
	return dns.RcodeSuccess, err
}
```

tpantelis commented:
The [second commit](https://github.com/submariner-io/lighthouse/pull/860/commits/0ca3deaf37dfa636f4ee4e09fc784befbfcb926c) is missing the `Signed-off-by`. Also the second overwrites the first so you might as well squash the commits.

Also these changes introduce [unit test failures](https://github.com/submariner-io/lighthouse/runs/7986090476?check_suite_focus=true).

blue-troy commented:
> The [second commit](https://github.com/submariner-io/lighthouse/pull/860/commits/0ca3deaf37dfa636f4ee4e09fc784befbfcb926c) is missing the `Signed-off-by`. Also the second overwrites the first so you might as well squash the commits.
> 
> Also these changes introduce [unit test failures](https://github.com/submariner-io/lighthouse/runs/7986090476?check_suite_focus=true).

thanks，[our unit test test the ServeDNS return code](https://github.com/submariner-io/lighthouse/blob/devel/coredns/plugin/handler_test.go#L1009) :

```go
	code, err := t.lh.ServeDNS(context.TODO(), rec, tc.Msg())

	Expect(code).Should(Equal(tc.Rcode))

	if tc.Rcode == dns.RcodeSuccess {
		Expect(err).To(Succeed())
		Expect(test.SortAndCheck(rec.Msg, tc)).To(Succeed())
	} else {
		Expect(err).To(HaveOccurred())
	}
```

but the Msg.code is more important , maybe we shoud check mgs.code and  ServeDNS's return error like below which coredns [kubernetes](https://github.com/coredns/coredns/blob/master/plugin/kubernetes/handler_case_test.go#L62) plugin do ? : 

```go
	_, err := k.ServeDNS(ctx, w, r)
		if err != tc.Error {
			t.Errorf("Test %d expected no error, got %v", i, err)
			return
		}
		if tc.Error != nil {
			continue
		}

		resp := w.Msg
		if resp == nil {
			t.Fatalf("Test %d, got nil message and no error for %q", i, r.Question[0].Name)
		}

		if err := test.SortAndCheck(resp, tc); err != nil {
			t.Error(err)
		}
```

blue-troy commented:
> our unit test test the ServeDNS return code

or we can simply return the dns.RcodeSuccess with empty ？

tpantelis commented on coredns/plugin/handler.go:
```diff
@@ -83,7 +83,10 @@ func (lh *Lighthouse) getDNSRecord(ctx context.Context, zone string, state *requ
+			if lh.Fall.Through(state.Name()) {
+				return plugin.NextOrFailure(lh.Name(), lh.Next, ctx, w, r) // nolint:wrapcheck // Let the caller wrap it.
+			}
+			return lh.emptyResponse(state, dns.RcodeNameError)
```
In looking at the [coredns server code](https://github.com/coredns/coredns/blob/fc7f3835d506960364b21cf62116a70eb8f4caf2/core/dnsserver/server.go#L263):

```Go
rcode, _ := h.pluginChain.ServeDNS(ctx, w, r)
if !plugin.ClientWrite(rcode) {
    errorFunc(s.Addr, w, r, rcode)
}
return
```

it assumes `ServeDNS` has written the response body unless the returned code indicates it hasn't been written as determined by `ClientWrite` (which is only the codes `RcodeServerFailure`, `RcodeRefused`, `RcodeFormatError` and `RcodeNotImplemented`), in which case it writes the response. 

So  in `lh.nextOrFailure`:

```Go
func (lh *Lighthouse) nextOrFailure(ctx context.Context, name string, w dns.ResponseWriter, r *dns.Msg, code int, err string) (int, error) {
    if lh.Fall.Through(name) {
        return plugin.NextOrFailure(lh.Name(), lh.Next, ctx, w, r) // nolint:wrapcheck // Let the caller wrap it.
    }

    return code, lh.error(err)
}
```

if there's no next plugin, it simply returns the code but if it isn't one of the `ClientWrite` codes, then no response is written, which is the case here with `RcodeNameError`. 

Rather than using `emptyResponse` here, I think we should modify `nextOrFailure` to return a response similarly. To reduce code duplication, we can define a general function:

```Go
func (lh *Lighthouse) writeResponse(state *request.Request, a *dns.Msg) (int, error) {
    a.Authoritative = true

    wErr := state.W.WriteMsg(a)
    if wErr != nil {
        log.Errorf("Failed to write message %#v: %v", a, wErr)
	return dns.RcodeServerFailure, lh.error("failed to write response")
    }

    return dns.RcodeSuccess, nil
}
```

Then `emptyResponse` becomes:

```Go
func (lh *Lighthouse) emptyResponse(state *request.Request) (int, error) {
    a := new(dns.Msg)
    a.SetReply(state.Req)

    return lh.writeResponse(state, a)
}
```

And `nextOrFailure` becomes:

```Go
func (lh *Lighthouse) nextOrFailure(ctx context.Context, name string, w dns.ResponseWriter, r *dns.Msg, code int) (int, error) {
    if lh.Fall.Through(name) {
        return plugin.NextOrFailure(lh.Name(), lh.Next, ctx, w, r) // nolint:wrapcheck // Let the caller wrap it.
    }

    a := new(dns.Msg)
    a.SetRcode(state.Req, rcode)

    return lh.writeResponse(state, a)
}
```

blue-troy replied:
good idea

blue-troy replied:
[in coredns code ](https://github.com/coredns/coredns/blob/master/plugin/plugin.go#L20) 
```go
	// Handler is like dns.Handler except ServeDNS may return an rcode
	// and/or error.
	//
	// If ServeDNS writes to the response body, it should return a status
	// code. CoreDNS assumes *no* reply has yet been written if the status
	// code is one of the following:
	//
	// * SERVFAIL (dns.RcodeServerFailure)
	//
	// * REFUSED (dns.RecodeRefused)
	//
	// * FORMERR (dns.RcodeFormatError)
	//
	// * NOTIMP (dns.RcodeNotImplemented)
```

so only this 4 code can not use writeResponse method

tpantelis replied:
No it doesn't mean we can't handle those codes - it just means we don't have to. But we might as well to simplify our code.

tpantelis commented:
@vthapar @aswinsuryan Can you review?

blue-troy committed:
Write the dns message response body when it is not a ServerFailure

Coredns assumes ServeDNS has written the response body unless
the returned code indicates it hasn't been written as determined
by ClientWrite (which is only the codes RcodeServerFailure,
RcodeRefused, RcodeFormatError and RcodeNotImplemented),
in which case it writes the response.

If there's no next plugin, it simply returns the code but if
it isn't one of the ClientWrite codes, then no response is written.

If no response,client will try to resend dns request multiple times,
which causes seconds delay.

Fixes #859
Signed-off-by: blue-troy <12729455+blue-troy@users.noreply.github.com>

tpantelis commented on coredns/plugin/handler.go:
```diff
@@ -162,10 +166,13 @@ func (lh *Lighthouse) error(str string) error {
 }
 
-func (lh *Lighthouse) nextOrFailure(ctx context.Context, name string, w dns.ResponseWriter, r *dns.Msg, code int, err string) (int, error) {
+func (lh *Lighthouse) nextOrFailure(ctx context.Context, name string, state *request.Request, r *dns.Msg, rcode int) (int, error) {
```
Since callers always pass `state.Name()`, we can eliminate the `name` parameter since we have the `state` instance.

blue-troy replied:
ok, it is solved

blue-troy committed:
Merge branch 'devel' into fix/not_found_records_slow

dfarrell07 commented:
@aswinsuryan to verify if this is actually user-facing, needs release note. - PR scrub

aswinsuryan commented:
> @aswinsuryan to verify if this is actually user-facing, needs release note. - PR scrub

user should see improvements when querying a record not available in "clusterset.local" , so can be added in release note.
